### ISSUE TITLE
AVT3 - check the screen reader for all guides

### ISSUE THREAD

**Issue by gkwan-ibm** [OpenLiberty/guides-common#491]
- Check general by Darren
- Check following by Danny: 
  - diagram
  - tables
  - commands
  - output areas
  - right pane
  - hotspots
  - os tabs
  - url links

**Comment by jj-zhang** [comment]
There are three steps to accessibility verification testing (AVT).  Each step should be done in order so that the same problems don’t get found again in later steps and makes the later step much easier if most of the issues have already been found and fixed.
AVT 1 = Automated testing - which is the accessibility checker. The newest version that was just released last month works in both FF and Chrome.  Yes, this checker will catch the majority of the issues and problems.
AVT 2 = Manual testing - these tests are one that require human intervention. Some of the “Potential Violations (PVs)” from automated testing might need a human user to check whether or not there is a real problem that needs to be fixed. Or, there are some other things that just needs a human eye to verify. For example, a picture requires alternative text to describe the image. However, if the picture is a Banana but the alt text says Apple, then that would be a problem which the automated tester wouldn’t be able to catch.
AVT 3 = Screen Reader testing - which is required and much be done with JAWS screen reader.  IBM has a site license for JAWS for testing purposes. If you can’t find the info about that, let me know and I can help you get it,


Note: Need to read https://www.ibm.com/able/guidelines/ci162/accessibility_checklist.html to find out what to test for the AVT 1, 2, and 3


STEPS
AVT 1
- download Dynamic assessment plugin (DAP) on google chrome browser here: (https://chrome.google.com/webstore/detail/dynamic-assessment-plugin/aahpafpbmmgednbflpalchnlbicaeppi)
- go to the URL you want to scan


Using Chrome DAP
Load web page which you would like to scan for accessibility issues
In the Chrome Menu, choose More Tools > Developer Tools
Select the "Elements" panel and then select the "DAP" panel in the right subpanel
Open URL chrome-extension://aahpafpbmmgednbflpalchnlbicaeppi/documentation/usingDAP.htm to access the Dynamic Assessment Plugin User Guide after installing DAP.
If your scan starts, but does not show progress or ever finish, you may need to enable the extension to access file URLs:
Load web page which you would like to scan for accessibility issues
In the Chrome Menu, choose More Tools > Extensions
Find the Dynamic Assessment Plugin extension. Enable Allow access to file URLs.


- You may need to get an authentication token for rules upload. This can be found here: https://aat.w3ibm.mybluemix.net/tools/auth/tools/dap.html?accountId=a4049770-b5e6-43ff-89a7-098b6b639208 


AVT 2 -> Check for the following:
Non-text Content: Any information conveyed by means other than text needs to be provided in text as well. The text should provide equivalent information. Examples of non-text content that convey information to users are images, graphs, media files, animations, CAPTCHA and audio alerts.
screen reader should read the alt text of the image, so make sure the alt text has enough explanation of the image
move the mouse over the image to see any popup text
Meaningful Sequence: An example of meaningful sequence is text in a two-column article. A user must read the lines of text in the first column sequentially, then move to the second column and do the same. If an assistive technology reads across both columns (e.g., reads the first line of the first column then the first line of the second column) before proceeding to the next line, the user will obviously not understand the meaning. Other potential failures of meaningful sequence can occur when using CSS, layout tables or white space to position content.
Screen reader should be able to read table content and read it by column instead of row
Input Purpose: Make sure input fields clearly indicate what the input field wants. 
Screen reader should clearly indicate what kind of input an input field wants
Use of Color: Color cannot be the only visual means of conveying information, because many users have deficiencies in color perception and cannot identify or distinguish colors. This checkpoint requires that you provide another non-color method for making the information available visually.
Check that any text using colour to convey a message also uses other means to convey the message such as with pattern, images, text, or text formatting. 
Contrast (Minimum): Content must provide enough contrast between text and its background to enable users with moderately low vision to read the text without contrast-enhancing technology. (Users requiring more contrast will often use an assistive technology or platform feature such as Windows High Contrast mode.) WCAG 2.0 makes a distinction between "large" text and small or body text.  Large text, which is defined as text that is at least 18 point or at least 14 point when bold, has a contrast requirement of 3:1.  Body text (less than 18 point and less than 14 point when bold) has a contrast requirement of 4.5:1.
Use https://color.a11y.com/ to check contrast issues
Resize text: Verify that text can be scaled up and down (can increase/decrease through the browser).
Scale up & down the size of the browser and verify that everything is readable and theres no lack of appropriate line wrapping, truncated, overlapped or obscured text, or missing scrollbars.
Content on Hover or Focus: Make sure all hover elements (eg. copy to clipboard) stay visible and dont disappear when hovered over and are accessible through the keyboard. 
Verify the copy to clipboard and any other hover element is read by the screen reader
Verify they don't disappear on mouse hover
Verify they are accessible by the keyboard
Keyboard: All functionality of the content is operable through a keyboard interface without requiring specific timings for individual keystrokes.
Verify you can go through the whole guide using only the keyboard
Page Titled: 
Web pages, non-web documents, and software have titles that describe topic or purpose
Link Purpose (In Context): The purpose of each link can be determined from the link text alone or from the link text together with its programmatically determined link context.
Screen reader can identify the purpose of a link from the link text alone or a alt text
 Language of Page: The default human language of Web pages, non-Web documents, or software can be programmatically determined (Through <Lang> element in HTML)
screen reader can determine was language webpage is in?
Error Suggestion: If an input error is automatically detected and suggestions for correction are known, then the suggestions are provided to the user.
Screen reader can detect any error messages thrown by bad inputs and etc.




TEST Results
Darren check 
the crash guides (only happen in Firefox, and Chrome is fine)
guide-getting-started
jpa-intro
cors
security-intro
docker
containerize
kubernetes-intro
cloud-ibm
cloud-openshift
cloud-aws
cloud-azure
cloud-google
the guides which only read the headers
No problem
Can use direction keys: up, down, left, right


Danny check to read
 diagram (reactive)
 tables (cors, jwt, security,...)
 commands
 output areas
 right pane
 hotspots
 os tabs (kubes)
 url links


diagram (having trouble)
tables (guide-cors)
commands (guide-microprofile-config)
output areas (guide-microprofile-fallback)
right pane (guide-microprofile-health)
hotspots (guide-microprofile-health
os tabs (guide-spring-boot)
url links (guide-microprofile-metrics)

**Comment by jj-zhang** [comment]
Darren check 
the crash guides (only happen in Firefox, and Chrome is fine)
guide-getting-started
jpa-intro
cors
security-intro
docker
containerize
kubernetes-intro
cloud-ibm
cloud-openshift
cloud-aws
cloud-azure
cloud-google
the guides which only read the headers
No problem
Can use direction keys: up, down, left, right

**Comment by DannyS03** [comment]
### Check to read the following: 
> - diagram 
> - tables 
> - commands
> - output areas
> - right pane
> - hotspots
> - os tabs 
> - url links

- Experiencing difficulties 
  - diagram (having trouble)
- Check was successful 
  - tables (guide-cors)
  - commands (guide-microprofile-config)
  - output areas (guide-microprofile-fallback)
  - right pane (guide-microprofile-health)
  - hotspots (guide-microprofile-health
  - os tabs (guide-spring-boot)
  - url links (guide-microprofile-metrics)
